File: go/word_search_big.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strings"
	"time"
)

func searchWordInFile(filePath string, word string) (float64, error) {
	file, err := os.Open(filePath)
	if err != nil {
		fmt.Println("Unable to open file:", filePath)
		return 0, errors.New("Unable to open file")
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	y := 0
	start := time.Now()

	for scanner.Scan() {
		y++
		if strings.Contains(scanner.Text(), word) {
			break
		}
	}

	duration := time.Since(start)

	//fmt.Printf("Time taken: %v milliseconds.\n", duration.Seconds()*1000)
	//fmt.Printf("%v,", duration.Seconds()*1000)
	return duration.Seconds() * 1000, nil
}

func saveExperimentToCSV(word string) {
	// Create or open a file for writing
	file, err := os.Create("go-results.csv")
	if err != nil {
		fmt.Println("Error creating file:", err)
		return
	}
	defer file.Close() // Ensure the file is closed after writing

	// HEADER
	_, err = fmt.Fprintf(file, "end,middle,start\n")
	if err != nil {
		fmt.Println("Error writing to file:", err)
		return
	}

	for i := 0; i < 100; i++ {
		endTime, _ := searchWordInFile("../text/big-end.txt", word)
		middleTime, _ := searchWordInFile("../text/big-middle.txt", word)
		startTime, _ := searchWordInFile("../text/big-start.txt", word)

		_, err = fmt.Fprintf(file, "%v,%v,%v\n", endTime, middleTime, startTime)
	}

	fmt.Println("Data successfully written to file.")
}

func main() {
	word := "pianissimo"
	saveExperimentToCSV(word)
}

File: rust/word_search_big.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader, BufWriter, Write};
use std::path::Path;
use std::time::Instant;

fn main() -> io::Result<()> {
    let search_word = "pianissimo";
    let file_locations = [
        "../text/big-end.txt",
        "../text/big-middle.txt",
        "../text/big-start.txt",
    ];

    let file = File::create("rust-results.csv")?;
    let mut file = BufWriter::new(file);
    writeln!(file, "end,middle,start")?;

    for _ in 0..100 {
        for (idx, &location) in file_locations.iter().enumerate() {
            let duration = search_word_in_file(location, search_word);

            if idx > 0 {
                write!(file, ",")?;
            }
            write!(file, "{:.4}", duration)?;
        }
        writeln!(file)?;
    }

    Ok(())
}

fn search_word_in_file(file_path: &str, search_word: &str) -> f64 {
    let path = Path::new(file_path);
    let start = Instant::now();

    let file = File::open(&path).unwrap();
    let reader = BufReader::new(file);

    for line in reader.lines() {
        if let Ok(line) = line {
            if line.contains(&search_word) {
                break;
            }
        }
    }
    start.elapsed().as_micros() as f64 / 1000.0
}